src/profilers.cpp

```diff
@@ -2,19 +2,375 @@
 // Copyright (C) CERN. See LICENSE for details.
 
 #include "profilers.hpp"
+#include "server/server.hpp"
 #include <cstdlib>
 #include <future>
 #include <iostream>
 #include <signal.h>
 #include <sys/wait.h>
 #include <fcntl.h>
 #include <nlohmann/json.hpp>
+#include <Poco/Net/StreamSocket.h>
+#include <boost/core/demangle.hpp>
+#include <boost/algorithm/string.hpp>
+#include <boost/program_options/parsers.hpp>
+
 
 #ifndef APERF_SCRIPT_PATH
 #define APERF_SCRIPT_PATH "."
 #endif
 
+
+
 namespace aperf {
+
+
+  /**
+    Constructs a MetricReader object.
+
+    @param metric_command  Metric command to be executed
+    @param metric_name     Alias for the metric command given by the user                
+    @param freq            The frequency at which the metric command will be "sampled"                   
+    @param regex           Regular expression provided by the user 
+                           to parse the output from the metric command
+    @param server_buffer   Size of the connection buffer to the server        
+  */
+
+  MetricReader::MetricReader(std::string metric_command, 
+                            std::string metric_name, 
+                            int freq,
+                            std::string regex,
+                            unsigned int server_buffer){
+    
+    this->metric_command = metric_command;
+    this->metric_name = metric_name;
+    this->freq = freq;
+    this->regex = regex;
+    this->server_buffer = server_buffer;
+    this->name = "MetricReader";
+
+  }
+
+  std::string MetricReader::get_name() {
+    return this->name;
+  }
+
+  void MetricReader::start(pid_t pid,
+                   ServerConnInstrs &connection_instrs,
+                   fs::path result_out,
+                   fs::path result_processed,
+                   bool capture_immediately) {
+
+    std::string instrs = connection_instrs.get_instructions(this->get_thread_count());
+
+    //sanitize metric name for log file name
+    std::regex invalidChars(R"([\/:*?"<>|\s]*)");
+
+    std::string sanitized_metric_name = std::regex_replace(this->metric_name, invalidChars, "_");
+
+    if (sanitized_metric_name.empty()) {
+        sanitized_metric_name = "metric_command_stderr.log";
+    }else{
+      sanitized_metric_name += "_stderr.log";
+    }
+
+    fs::path stderr_metric_command;
+    stderr_metric_command = result_out / sanitized_metric_name;
+
+    const int ERROR_STDERR = 201;
+    const int ERROR_STDOUT_DUP2 = 202;
+    const int ERROR_STDERR_DUP2 = 203;
+    const int ERROR_NO_NUMBER_REGEX = 204;
+    const int ERROR_TOO_MANY_NUMBERS_REGEX = 205;
+    const int ERROR_CONVERSION_TO_FLOAT = 206;
+    const int ERROR_PARSING_CONNECTION_INSTRS = 207;
+    const int ERROR_USER_REGEX_MATCH = 208;
+    const int ERROR_PIPE_METRIC_EXEC = 209;
+    
+    
+    std::future<int> metric_profiler = std::async(std::launch::async, [=]() -> int
```

Comment:
Makes sense. I will try to rearrange the code a bit to merge the async functions.  I realized that I forgot to put return instead of std::exit in multiple places :sweat_smile:

Reply:
Should be done. I merged metric_profiler and process. Also, I made a function to handle errors .